src/vm/eventpipebuffer.cpp

```diff
@@ -144,65 +149,49 @@ EventPipeEventInstance *EventPipeBuffer::GetNext(EventPipeEventInstance *pEvent,
     CONTRACTL_END;
 
     EventPipeEventInstance *pNextInstance = NULL;
-    // If input is NULL, return the first event if there is one.
-    if (pEvent == NULL)
-    {
-        // If this buffer contains an event, select it.
-        BYTE *pFirstAlignedInstance = GetNextAlignedAddress(m_pBuffer);
-        if (m_pCurrent > pFirstAlignedInstance)
-        {
-            pNextInstance = (EventPipeEventInstance *)pFirstAlignedInstance;
-        }
-        else
-        {
-            return NULL;
-        }
-    }
-    else
+
+    // If m_pCurrentReadEvent is NULL we've reached the end of the events
+    if (m_pCurrentReadEvent != NULL)
     {
         // Confirm that pEvent is within the used range of the buffer.
-        if (((BYTE *)pEvent < m_pBuffer) || ((BYTE *)pEvent >= m_pCurrent))
+        if (((BYTE*)m_pCurrentReadEvent < m_pBuffer) || ((BYTE*)m_pCurrentReadEvent >= m_pCurrent))
         {
             _ASSERT(!"Input pointer is out of range.");
-            return NULL;
+            m_pCurrentReadEvent = NULL;
         }
 
-        if (pEvent->GetData())
+        if (m_pCurrentReadEvent->GetData())
```

Comment:
In the case you set this pointer to NULL in if block above, this will AV.

Reply:
Would it make more sense to create a
```cpp
EventPipeEventInstance *GetNextAlignedAddress(EventPipeEventInstance *pEventInstance)
{
    const auto Offset = m_pCurrentReadEvent->GetDataLength() ? m_pCurrentReadEvent->GetDataLength() : 1;
    return const_cast<BYTE *>(m_pCurrentReadEvent->GetData() + Offset);
}
```

Reply:
AV - yep that is bad : ) The code shouldn't ever go down that path but no reason to live so dangerously.

Refactoring-wise I'm going to leave this as-is unless you feel strongly about it. Certainly it could also be written with the ternary operator, but the line would get pretty long or you would need to split it into multiple lines. The expression above isn't quite the right calculation, I think it would need to be something like:

```
EventPipeEventInstance *GetNextAlignedAddress(EventPipeEventInstance *pEventInstance)
{
    const BYTE* pUnalignedAddress = m_pCurrentReadEvent->GetDataLength() ? 
        const_cast<BYTE*>(m_pCurrentReadEvent->GetData() + m_pCurrentReadEvent->GetDataLength()) : 
        const_cast<BYTE*>(m_pCurrentReadEvent + 1);
    return (EventPipeInstance*)GetNextAlignedAddress(pUnalignedAddress);
}
```

Extracting that out as a function might make the code in this function a little easier to follow, but IMO it felt like too small of an improvement not to give deference to the pre-existing code structure.